### PR TITLE
RFC: Hide zlib and expat libs from the user of Poco libraries

### PR DESCRIPTION
**Pull request by teksturi**
Seperete zlib and expat so they are only expose internally. There should not be reason for user to use these.

This makes https://github.com/pocoproject/poco/issues/4358 easier to implement as all external libraries are not exposed to end user.

### CODE REVIEW

**Comment by teksturi** [comment]
Please say if we example still want to expose ParserEngine. I will try to make necessary changes for that. I do not remember exactly if that would be easy as this has been quite long in my POC box.

File: Foundation/include/Poco/DeflatingStream.h [hunk 1]
```diff
@@ -104,7 +101,7 @@ class Foundation_API DeflatingIOS: public virtual std::ios
 	/// order of the stream buffer and base classes.
 {
 public:
-	DeflatingIOS(std::ostream& ostr, DeflatingStreamBuf::StreamType type = DeflatingStreamBuf::STREAM_ZLIB, int level = Z_DEFAULT_COMPRESSION);
+	DeflatingIOS(std::ostream& ostr, DeflatingStreamBuf::StreamType type = DeflatingStreamBuf::STREAM_ZLIB, int level = -1);
```
Comment (andrewauclair) [inline Foundation/include/Poco/DeflatingStream.h:104]: Going from `Z_DEFAULT_COMPRESSION` to just `-1` is a loss of info for users. Can we add a `POCO_DEFAULT_COMPRESSION` const for this?

File: Foundation/include/Poco/InflatingStream.h [hunk 2]
```diff
@@ -22,11 +22,8 @@
 #include "Poco/BufferedStreamBuf.h"
 #include <istream>
 #include <ostream>
-#if defined(POCO_UNBUNDLED)
-#include <zlib.h>
-#else
-#include "Poco/zlib.h"
-#endif
+
+struct z_stream_s;
```
Comment (andrewauclair) [inline Foundation/include/Poco/InflatingStream.h:26]: Don't think this really needs to be here. Just do `struct z_stream_s* _pZstr;` in the class (unless I'm not seeing another usage here).

File: Foundation/src/DeflatingStream.cpp [hunk 3]
```diff
@@ -25,29 +31,17 @@ DeflatingStreamBuf::DeflatingStreamBuf(std::istream& istr, StreamType type, int
...
-	_zstr.data_type = 0;
-	_zstr.adler     = 0;
-	_zstr.reserved  = 0;
-
-	_buffer = new char[DEFLATE_BUFFER_SIZE];
-
-	int rc = deflateInit2(&_zstr, level, Z_DEFLATED, 15 + (type == STREAM_GZIP ? 16 : 0), 8, Z_DEFAULT_STRATEGY);
+	std::unique_ptr<char[]> buffer(new char[DEFLATE_BUFFER_SIZE]);
+
+	std::unique_ptr<z_stream> pZstr = std::make_unique<z_stream>(z_stream{});
```
Comment (matejk) [inline Foundation/src/DeflatingStream.cpp:36]: The argument `z_stream{}` to `make_unique` seems wrong. Parameters for the ctor shall be passed there, probably the ones used before to initialize `_zstr`.

File: Foundation/testsuite/src/ZLibTest.cpp [hunk 4]
```diff
@@ -17,6 +17,12 @@
 #include "Poco/StreamCopier.h"
 #include "Poco/Buffer.h"
 #include <sstream>
+#if defined(POCO_UNBUNDLED)
+#include <zlib.h>
+#else
+// Quirk before we move zlib to external libs.
+#include "../../src/zlib.h"
```
Comment (matejk) [inline Foundation/testsuite/src/ZLibTest.cpp:24]: The path to zlih.h shall be added CMake's include path instead.

File: Foundation/src/DeflatingStream.cpp (hunk 3 above)
Comment (andrewauclair) [reply Foundation/src/DeflatingStream.cpp:36]: Assuming `Z_NULL` isn't something special and is just a `nullptr` (I would have to check), then this does the same thing. `z_stream{}` value-initializes a `z_stream` and `std::make_unique<z_stream>(z_stream{})` calls the move constructor.

Comment (matejk) [reply Foundation/src/DeflatingStream.cpp:36]: Thanks, I was not sure if default init of integers sets the values to 0.

Comment (andrewauclair) [reply Foundation/src/DeflatingStream.cpp:36]: Default init does not, but value init does. `int i;` is uninitialized, but `int i{};` is 0.

**Comment by matejk** [comment]
@teksturi , any updates on this?

File: Foundation/include/Poco/DeflatingStream.h (hunk 1 above)
Comment (matejk) [reply Foundation/include/Poco/DeflatingStream.h:104]: There is an issue with other (non-default) compression levels whenever the user decided to go with non-default.

Shall Poco maintain a copy of compression values that are identical with the ones in zlib?

@andrewauclair , @teksturi

**Comment by matejk** [comment]
Replaced with #4724.

**Pull request closed without merging**